lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -1758,3 +1793,56 @@ def _relabel_configs(self) -> list:
             if self._relabel_instance
             else []
         )
+
+
+class _TransformationProvider(Object):
+    """Enables runtime transformations of promql expressions using promql-transform."""
+
+    def __init__(self, charm: CharmBase, relation_name: str = DEFAULT_RELATION_NAME):
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._path: PosixPath = None
+        arch = platform.processor()
+        self._arch = "amd64" if arch == "x86_64" else arch
+        self._valid_architectures = ["amd64", "arm64"]
+
+        if self._arch not in self._valid_architectures:
+            raise UnsupportedTransformArchError(self._arch)
+
+        self._fetch_transformer_path(self._arch)
+
+    def _fetch_transformer_path(self, arch: str):
+        if self._path is not None:
+            return
+
+        res = "promql-transform-{}".format(arch)
+        try:
+            self._path = self._charm.model.resources.fetch(res)
+        except NameError:
+            raise MissingTransformResourceError(res)
+
+    def apply_label_matcher(self, expr: str, additional_label_filters: dict) -> str:
+        if self._path is None:
+            logger.debug(
+                "No `promql-transform` available, skipping the application of the label matchers."
+            )
+            return expr
+
+        os.chmod(self._path, 0o755)
+        args = [
+            '--label-matcher {}="{}"'.format(key, value)
+            for key, value in additional_label_filters.items()
+        ]
+
+        po = subprocess.Popen(
+            args="{} {} {}".format(self._path, " ".join(args), "'{}'".format(expr)),
+            shell=True,
+            stdout=subprocess.PIPE,
+        )
+
+        return_code = po.wait()
+
+        if return_code:
+            raise ValueError("An error occurred while invoking `promql-transform`.")
```

Comment:
since Popen [raises exceptions](https://docs.python.org/3.8/library/subprocess.html#exceptions), maybe this should be inside a try-except. `os.chmod` too.

Reply:
I'll add it for Popen. Would you mind pointing me at something explaining what kind of exceptions chmod raises? Can't find anything on python.org (or anywhere else)

Reply:
I triggered them manually myself...
`FileNotFoundError`, `PermissionError` are the most obvious to me. Maybe `OSError` too? With these kind of funcs I see no reason not to have a bare exception (`except Exception as e`).